plugins/inputs/windows_event_log/wineventlog/wineventlog_test.go

```diff
@@ -0,0 +1,163 @@
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	// Without a sleep we do not read any events!!!
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadBadSource will verify that we cannot read events written by an
+// unregistered event log source.
+func TestReadBadSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 0, "[Application] [ERROR] [888] [CWA_UnitTest222] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadWithBothSources will verify we can read events written by a
+// registered event log source, even if the batch contains events from an
+// unregistered source too.
+func TestReadWithBothSources(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// seekToEnd skips past all current events in the event log.
+func seekToEnd(t *testing.T, elog *windowsEventLog) {
+	// loop until we stop getting records.
+	numRetrieved := 0
+	for {
+		records := elog.read()
+		//assert.Equal(t, nil, err)
+		t.Logf("seekToEnd() current count %v", len(records))
+		numRetrieved += len(records)
+		if len(records) == 0 {
+			break
+		}
+	}
+	t.Logf("seekToEnd() total %v", numRetrieved)
+}
+
+// writeEvents writes msgCount number of events to the Application event log.
+// Optionally register the given logSrc.
+// Fail the test if an error occurs.
+func writeEvents(t *testing.T, msgCount int, doRegister bool, logSrc string, eventId uint32) {
+	if doRegister {
+		// Expected to fail if unit test previously ran and installed the event src.
+		_ = eventlog.InstallAsEventCreate(logSrc, eventlog.Info|eventlog.Warning|eventlog.Error)
+	}
+	wlog, err := eventlog.Open(logSrc)
+	assert.NoError(t, err)
+	for i := 0; i < msgCount; i++ {
+		msg := fmt.Sprintf("CWA_UnitTest event msg %v", i)
+		wlog.Error(eventId, msg)
+	}
+	err = wlog.Close()
+	assert.NoError(t, err)
+}
+
+// checkEvents reads all events (since last read) then checks each one for the
+// given substr. Verify count == msgCount.
+func checkEvents(t *testing.T, elog *windowsEventLog, msgCount int, substr string) {
+	// Get all new records.
+	var records []*windowsEventLogRecord
+	for {
```

Comment:
Is there a possibility for this to run indefinitely? I see there's the `break` when the records we read is empty, but I feel like this is an easy thing to catch that could save a testing headache later on.

Reply:
If the VM had a never ending stream of events getting created and it outpaced our ability to read events then yes it could be infinite.
I will add a max loop count to it just to be safe.